modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/lifecycle/DefaultArtifactDisposalContext.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.lifecycle;
+
+import static java.lang.Thread.getAllStackTraces;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+import org.mule.runtime.core.internal.util.CompositeClassLoader;
+import org.mule.runtime.module.artifact.api.classloader.ArtifactClassLoader;
+import org.mule.sdk.api.artifact.lifecycle.ArtifactDisposalContext;
+import org.mule.sdk.api.artifact.lifecycle.ArtifactLifecycleListener;
+
+import java.util.stream.Stream;
+
+import org.slf4j.Logger;
+
+/**
+ * Default implementation of {@link ArtifactDisposalContext}.
+ */
+public class DefaultArtifactDisposalContext implements ArtifactDisposalContext {
+
+  private static final Logger LOGGER = getLogger(DefaultArtifactDisposalContext.class);
+
+  private final ArtifactClassLoader artifactClassLoader;
+  private final ArtifactClassLoader extensionClassLoader;
+
+  /**
+   * Creates an {@link ArtifactDisposalContext} from the given {@code artifactClassLoader} and {@code extensionClassLoader} pair.
+   *
+   * @param artifactClassLoader  the {@link ArtifactClassLoader} of the artifact being disposed.
+   * @param extensionClassLoader the {@link ArtifactClassLoader} of the extension the {@link ArtifactLifecycleListener} is
+   *                             associated with.
+   */
+  public DefaultArtifactDisposalContext(ArtifactClassLoader artifactClassLoader, ArtifactClassLoader extensionClassLoader) {
+    this.artifactClassLoader = artifactClassLoader;
+    this.extensionClassLoader = extensionClassLoader;
+  }
+
+  @Override
+  public ClassLoader getExtensionClassLoader() {
+    return extensionClassLoader.getClassLoader();
+  }
+
+  @Override
+  public ClassLoader getArtifactClassLoader() {
+    return artifactClassLoader.getClassLoader();
+  }
+
+  @Override
+  public Stream<Thread> getArtifactOwnedThreads() {
+    try {
+      // This may be expensive because it needs to build all the stack traces which we don't actually need.
+      // However, other alternatives like Thread#enumerate would require us to traverse up to the root Thread Group in order
+      // to actually get all threads.
+      // Because this is only executed on disposal, we prefer readability to marginal performance gains.
+      return getAllStackTraces().keySet().stream().filter(this::isArtifactOwnedThread);
```

Comment:
This is inspired on what the [IBM MQ Releaser does](https://github.com/mulesoft/mule/blob/451f70e349635c5833e9d2444de999e70385ba95/modules/artifact/src/main/java/org/mule/module/artifact/classloader/IBMMQResourceReleaser.java#L303).